### PR TITLE
Verification code

### PR DESCRIPTION
**Pull request by Yixin-Hu**
1. add unit test for edge splitting
2. add unit test for tuple switch operation
3. add assertions in tuple functions
4. add Tuple::init_from_face()
5. complete Tuple::get_conn_tets()
6. add Tuple::check_validity() for check validity of tuple
7. add check_mesh_connectivity_validity() for checking validity of mesh connectivity
8. move edge splitting, edge collapsing connectivity functions into separate .cpp fils

### CODE REVIEW

File: src/Tuple.cpp [hunk 1]
```diff
@@ -31,42 +59,52 @@ TetMesh::Tuple TetMesh::Tuple::init_from_tet(const TetMesh& m, int tid){
...
+    assert(!n12_t_ids.empty());
+
     int tid = *std::min_element(n12_t_ids.begin(), n12_t_ids.end());
-    for(int j=0;j<6;j++){
+    for (int j = 0; j < 6; j++) {
         int tmp_v1_id = m.m_tet_connectivity[tid][m_local_edges[j][0]];
         int tmp_v2_id = m.m_tet_connectivity[tid][m_local_edges[j][1]];
-        if(tmp_v1_id == v1_id && tmp_v2_id == v2_id
-            ||tmp_v1_id == v2_id && tmp_v2_id == v1_id)
+        if (tmp_v1_id == v1_id && tmp_v2_id == v2_id || tmp_v1_id == v2_id && tmp_v2_id == v1_id)
```
Comment (jiangzhongshi) [inline src/Tuple.cpp:107]: I would recommend add parenthesize for readability.
```suggestion
        if ((tmp_v1_id == v1_id && tmp_v2_id == v2_id) || (tmp_v1_id == v2_id && tmp_v2_id == v1_id))
```

File: src/Tuple.cpp [hunk 2]
```diff
@@ -100,30 +140,42 @@ size_t TetMesh::Tuple::fid(const TetMesh& m) const
...
+size_t TetMesh::Tuple::tid() const
+{
+    return m_tid;
+}
 
 TetMesh::Tuple TetMesh::Tuple::switch_vertex(const TetMesh& m) const
 {
     Tuple loc = *this;
+//    std::cout<<"switch_vertex"<<std::endl;
+//    std::cout<<loc.m_vid<<" "<<loc.m_eid<<" "<<loc.m_fid<<" "<<loc.m_tid<<std::endl;
```
Comment (jiangzhongshi) [inline src/Tuple.cpp:152]:
```suggestion
```

File: src/Tuple.cpp [hunk 3]
```diff
@@ -100,30 +140,42 @@ size_t TetMesh::Tuple::fid(const TetMesh& m) const
...
+
     int l_vid1 = m_local_edges[m_eid][0];
     int l_vid2 = m_local_edges[m_eid][1];
-    loc.m_vid = m.m_tet_connectivity[m_tid][l_vid1] == m_vid
-                    ? m.m_tet_connectivity[m_tid][l_vid2]
-                    : m.m_tet_connectivity[m_tid][l_vid1];
+    loc.m_vid = m.m_tet_connectivity[m_tid][l_vid1] == m_vid ? m.m_tet_connectivity[m_tid][l_vid2]
+                                                             : m.m_tet_connectivity[m_tid][l_vid1];
+    assert(loc.m_vid >= 0 && loc.m_vid < m.m_vertex_connectivity.size());
+//    std::cout<<"->"<<loc.m_vid<<" "<<loc.m_eid<<" "<<loc.m_fid<<" "<<loc.m_tid<<std::endl;
```
Comment (jiangzhongshi) [inline src/Tuple.cpp:159]:
```suggestion
```

File: src/Tuple.cpp [hunk 4]
```diff
@@ -100,30 +140,42 @@ size_t TetMesh::Tuple::fid(const TetMesh& m) const
...
 
     return loc;
 } // along edge
 
 TetMesh::Tuple TetMesh::Tuple::switch_edge(const TetMesh& m) const
 {
+//    std::cout<<"switch_edge"<<std::endl;
```
Comment (jiangzhongshi) [inline src/Tuple.cpp:166]:
```suggestion
```

**Review by jiangzhongshi — APPROVED** [review]
